qiime2/metadata/metadata.py

```diff
@@ -353,7 +324,49 @@ def column_count(self):
         """
         return len(self._columns)
 
-    def __init__(self, dataframe):
+    @property
+    def artifacts(self):
+        """Artifacts that are the source of the metadata.
+
+        This property is read-only.
+
+        Returns
+        -------
+        tuple of qiime2.Artifact
+            Source artifacts of the metadata.
+
+        """
+        if self._source_artifact is not None:
+            return (self._source_artifact,)
+
+        artifacts = set()
+        for source in self._column_sources.values():
+            if source is not None:
+                artifacts.update(artifact for artifact in source.artifacts)
+        return tuple(artifacts)
+
+    @property
+    def source_artifact(self):
+        """Artifact that is the source of the metadata.
+
+        This property is read-only.
+
+        Returns
+        -------
+        qiime2.Artifact
+            Source artifact of the metadata.
+
+        """
+        return self.source_artifact
```

Comment:
This appears to be a circular reference, should it be the following, instead?

```suggestion
        return self._source_artifact
```